### PR TITLE
Examples: GPGPU Water Port

### PR DESCRIPTION
**Pull request by cmhhelgeson**
Related issue: #XXXX

**Description**

Port of the WebGL GPGPU Water Sample. 

EDIT 1: Sphere dynamics have been modified to leverage to used the positionNode shader over the CPU. 

EDIT 2: Original WebGL sample slightly adjusted to demonstrate WebGL vs. WebGPU performance differential.

https://github.com/user-attachments/assets/d567f94b-52b0-40a2-add8-1f771f0a1b0f

### CODE REVIEW

**Comment by cmhhelgeson** [comment]
Kind of stumped by the WebGL Backend behavior here:

https://github.com/user-attachments/assets/afbc0a43-c864-4e08-8e14-d0bd40e9c115

**Comment by cmhhelgeson** [comment]
Think I understand the issue with the WebGL backend, it seems like the conversion to GLSL does not yet account for accessing the same buffer multiple times. For instance, this TSL code

```typescript
const { viscosity, mousePos, mouseSize } = effectController;

const height = heightStorage.element( instanceIndex );
const prevHeight = prevHeightStorage.element( instanceIndex );

const { north, south, east, west } = getNeighborValuesTSL( instanceIndex, heightStorage );

const neighborHeight = north.add( south ).add( east ).add( west );
neighborHeight.mulAssign( 0.5 );
neighborHeight.subAssign( prevHeight );

const newHeight = neighborHeight.mul( viscosity );
```

Generates this GLSL output:

```glsl
float nodeVar0;
float nodeVar1;

// transforms
nodeVarying0 = nodeAttribute0;
nodeVarying1 = nodeAttribute1;
	
// flow code

nodeVar0 = ( ( ( nodeVarying0 + nodeVarying0 ) + nodeVarying0 ) + nodeVarying0 );
nodeVar0 = ( nodeVar0 * 0.5 );
nodeVar0 = ( nodeVar0 - nodeVarying1 );
nodeVar1 = ( nodeVar0 * c_viscosity );

```
The first line of code corresponds with the getNeighborValuesTSL call but only the same height value it already read in.

Equivalent WGSL output for comparison
```wgsl

// system
instanceIndex = id.x + id.y * numWorkgroups.x * u32(64) + id.z * numWorkgroups.x * numWorkgroups.y * u32(64);

// vars
	
var nodeVar0 : u32;
var nodeVar1 : u32;
var nodeVar2 : f32;
var nodeVar3 : f32;


// flow code

nodeVar0 = ( instanceIndex / 128u );
nodeVar1 = ( instanceIndex % 128u );
nodeVar2 = ( ( ( NodeBuffer_476.Height[ ( ( min( ( nodeVar0 + u32( 1.0 ) ), ( 128u - u32( 1.0 ) ) ) * 128u ) + nodeVar1 ) ] + 
     NodeBuffer_476.Height[ ( ( max( u32( 0.0 ), ( nodeVar0 - u32( 1.0 ) ) ) * 128u ) + nodeVar1 ) ] ) + NodeBuffer_476.Height[ ( ( 
     nodeVar0 * 128u ) + min( ( nodeVar1 + u32( 1.0 ) ), ( 128u - u32( 1.0 ) ) ) ) ] ) + NodeBuffer_476.Height[ ( ( nodeVar0 * 128u ) 
     + max( u32( 0.0 ), ( nodeVar1 - u32( 1.0 ) ) ) ) ] );
nodeVar2 = ( nodeVar2 * 0.5 );
nodeVar2 = ( nodeVar2 - NodeBuffer_477.PrevHeight[ instanceIndex ] );
nodeVar3 = ( nodeVar2 * object.viscosity );
```

**Comment by mrdoob** [comment]
@RenaudRohlinger @sunag

**Comment by cmhhelgeson** [comment]
With storage objects, the GLSL Node Builder accesses the PBO, but not in the right place.

```glsl
// Data Layout
layout( std140 ) uniform compute_object {
	float c_viscosity;
	vec2 c_mousePos;
	float c_mouseSize;
};
uniform highp sampler2D Height;
uniform highp sampler2D PrevHeight;

// varyings
out float nodeVarying0;
out float nodeVarying1;

// attributes
layout( location = 0 ) in float nodeAttribute2;
layout( location = 1 ) in float nodeAttribute6;

// code 

nodeVar0 = ( ( ( nodeVarying0 + nodeVarying0 ) + nodeVarying0 ) + nodeVarying0 );
nodeVar0 = ( nodeVar0 * 0.5 );
nodeVar1Size = uint( textureSize( nodeUniform1, 0 ).x );

nodeVar1 = vec4(texelFetch( nodeUniform1, ivec2(uint( gl_InstanceID ) % nodeVar1Size, uint( gl_InstanceID ) / nodeVar1Size), 0 )).x;

nodeVar0 = ( nodeVar0 - nodeVar1 );
nodeVar3 = ( nodeVar0 * c_viscosity );
nodeVar3 = ( nodeVar3 + ( ( cos( clamp( ( ( length( ( ( ( vec2( ( float( ( uint( gl_InstanceID ) % uint( 128.0 ) ) ) * 0.0078125 ), ( float( ( uint( gl_InstanceID ) / uint( 128.0 ) ) ) * 0.0078125 ) ) - vec2( 0.5, 0.5 ) ) * vec2( 512.0 ) ) - c_mousePos ) ) * 3.141592653589793 ) / c_mouseSize ), 0.0, 3.141592653589793 ) ) + 1.0 ) * 0.28 ) );
nodeVar4Size = uint( textureSize( Height, 0 ).x );

nodeVar4 = vec4(texelFetch( Height, ivec2(uint( gl_InstanceID ) % nodeVar4Size, uint( gl_InstanceID ) / nodeVar4Size), 0 )).x;
nodeVarying1 = nodeVar4;
nodeVarying0 = nodeVar3;

```

File: examples/webgpu_compute_water.html [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+						prevHeightArray[ p ] = height;
+
+						p ++;
+
+					}
+
+				}
+
+				const heightBufferAttribute = new THREE.StorageBufferAttribute( heightArray, 1 );
+				const prevHeightBufferAttribute = new THREE.StorageBufferAttribute( prevHeightArray, 1 );
```
Comment (sunag) [inline examples/webgpu_compute_water.html:130]: `StorageInstancedBufferAttribute`

File: examples/webgpu_compute_water.html [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+				waterMaterial.lights = true;
+				waterMaterial.colorNode = color( 0x0040C0 );
+				waterMaterial.specularNode = color( 0x111111 );
+				waterMaterial.shininess = Math.max( 50, 1e-4 );
+				waterMaterial.positionNode = Fn( () => {
+
+					// To correct the lighting as our mesh undulates, we have to reassign the normals in the position shader.
+					const { normalX, normalY } = getNormalsFromHeightTSL( vertexIndex, heightRead );
+
+					varyingProperty( 'vec3', 'v_normalView' ).assign( modelNormalMatrix.mul( vec3( normalX, negate( normalY ), 1.0 ) ) );
```
Comment (sunag) [inline examples/webgpu_compute_water.html:260]: TSL now have a function for this: `transformNormalToView( vec3( normalX, negate( normalY ), 1.0 ) )`

Comment (cmhhelgeson) [reply examples/webgpu_compute_water.html:260]: This seems to break the lighting when I test it. Aren't modelNormalMatrix and modelNormalViewMatrix distinct? And if they aren't, should the name of the node be updated to reflect that?

**Pull request merged**